src/compiler/grammar/Shexl.g4

```diff
@@ -0,0 +1,276 @@
+ ;
+
+// A triple constraint is the abstraction of the well known
+// eachOf of ShEx. In our case as we don't have the eachOneOf
+// we don't need to expand this rule. Therefore the triple
+// constraint is defined as a property, a node constraint and
+// the cardinality. If no cardilaity is pressent the default one
+// will be the [0,n).
+triple_constraint
+ : prefix_invocation node_constraint cardinality?
+ ;
+
+// A prefix invocation is the invocation to something declared
+// in the prefix. For example when we declare a property name
+// in a triple constraint we invoke the property associated to
+// the declared prefix. And the same with Datatypes.
+prefix_invocation
+ : prefix=LABEL? ':' property=LABEL
+ | A_KW     // keywork `a` alias for rdf:type
+ ;
+
+// Describe the allow values of a node. It can be:
+//  - anything.
+//  - a datatype. Notice that a datatype actually is a prefix
+//      invocation.
+//  - a node kind.
+//  - a value set.
+//  - or a shape reference.
+node_constraint
+ : '.'                  // Anything.
+ | prefix_invocation    // Datatype.
+ | LITERAL_KW           // Node kind.
+ | IRI_KW
+ | BNODE_KW
+ | NON_LITERAL_KW
+ | '[' value_set_type* ']'  // Value set.
+ | shape_invocation     // Shape reference.
+ ;
+
+// Represents the possible values that a value set might contain.
+value_set_type
+ : prefix_invocation    // Prefix invocation.
+ | shape_invocation     // Shape invocation.
+ | STRING_LITERAL       // String literal.
+ | REAL_LITERAL         // Real literal.
+ ;
+
+// As seen previously a triple constraint is formed by a property, a
+// node constraint and a cardinality. The cardinality might take different
+// values depending on what does it mean:
+//  - `*` -> any number of repetitions, 0 or more. If no cardinality
+//           is set this one is de default one. [0,n)
+//  - `+` -> any number of repetitions but at least 1. [1,n)
+//  - `?` -> none or one appearance. Also known as optional. [0,1]
+//  - `{N}` -> exactly N repetitions [N,N].
+//  - `{N,M}` -> a minimum of N and a maximum of M repetitions. [N,M]
+//  - `{N, }` -> any number of repetitions but a minimum of N. [N,m)
+cardinality
+ : '*'
+ | '+'
+ | '?'
+ | '{' INT_LITERAL '}'
+ | '{' INT_LITERAL ',' INT_LITERAL '}'
+ | '{' INT_LITERAL ',''}'
+ ;
+
+// **********************************
+// TOKENS
+// **********************************
+
+// Keyworkds of the language.
+PREFIX_KW       :   'PREFIX'        ;
+BASE_KW         :   'BASE'          ;
+IMPORT_KW       :   'IMPORT'        ;
+START_KW        :   'START'         ;
+AND_KW          :   'AND'           ;
+OR_KW           :   'OR'            ;
+NOT_KW          :   'NOT'           ;
+IRI_KW          :   'IRI'           ;
+LITERAL_KW      :   'LITERAL'       ;
+BNODE_KW        :   'BNODE'         ;
+NON_LITERAL_KW  :   'NONLITERAL'    ;
+CLOSED_KW       :   'CLOSED'        ;
+A_KW            :   'a'             ;
+
+
+LABEL
+ : [a-zA-Z_][a-zA-Z0-9_-·]*
```

Comment:
I think labels can start with a number. Everything else seems perfect to me.

Reply:
Yeah that’s something tricky, I’m not quite sure that Turtle nor SPARQL allow that on their specification, I’m being cautious before allowing more things than the ones allowed. As the syntax is intended to be a subset allow less things is not a problem, we can always increase the pattern of the lexeme at any time.